### PR TITLE
COO matrix

### PR DESCRIPTION
**Pull request by unnonouno**
I implemented COO matrix.
Please merge #226 first.
related to #36

### CODE REVIEW

File: cupy/sparse/coo.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+    def transpose(self, axes=None, copy=False):
+        """Returns a transpose matrix.
+
+        Args:
+            axes: This option is not supported.
+            copy (bool): If ``True``, a returned matrix shares no data.
+                Otherwise, it shared data arrays as much as possible.
+
+        """
```
Comment (takagi) [inline cupy/sparse/coo.py]: Please add what it returns.

File: cupy/sparse/coo.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    def tocsr(self, copy=False):
+        """Converts the matrix to Compressed Sparse Row format.
+
+        Args:
+            copy (bool): If ``False``, it shares data arrays as much as
+                possible. Actually this option is ignored because all
+                arrays in a matrix cannot be shared in coo to csr conversion.
+
+        Returns:
+            cupy.sparse.csr_matrix: Converte matrix.
```
Comment (takagi) [inline cupy/sparse/coo.py]: typo "Converted"

File: cupy/sparse/coo.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    def tocsc(self, copy=False):
+        """Converts the matrix to Compressed Sparse Column format.
+
+        Args:
+            copy (bool): If ``False``, it shares data arrays as much as
+                possible. Actually this option is ignored because all
+                arrays in a matrix cannot be shared in coo to csc conversion.
+
+        Returns:
+            cupy.sparse.csc_matrix: Converte matrix.
```
Comment (takagi) [inline cupy/sparse/coo.py]: typo "Converted"

File: cupy/sparse/coo.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+    def tocoo(self, copy=False):
+        """Converts the matrix to COOdinate format.
+
+        Args:
+            copy (bool): If ``False``, it shares data arrays as much as
+                possible.
+
+        Returns:
+            cupy.sparse.coo_matrix: Converte matrix.
```
Comment (takagi) [inline cupy/sparse/coo.py]: typo "Converted"

File: cupy/sparse/coo.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        if dtype is None:
+            dtype = data.dtype
+        else:
+            dtype = numpy.dtype(dtype)
+
+        if dtype != 'f' and dtype != 'd':
+            raise ValueError('Only float32 and float64 are supported')
+
+        data = data.astype(dtype, copy=copy)
+        row = row.astype('i', copy=copy)
```
Comment (takagi) [inline cupy/sparse/coo.py]: `numpy.intp` would be better.

File: cupy/sparse/coo.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        col = col.astype('i', copy=copy)
```
Comment (takagi) [inline cupy/sparse/coo.py]: `numpy.intp` would be better.

File: tests/cupy_tests/sparse_tests/test_coo.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+        ]
+        cupy.testing.assert_allclose(m, expect)
+
+
+@testing.parameterize(*testing.product({
+    'dtype': [numpy.float32, numpy.float64],
+}))
+@unittest.skipUnless(scipy_available, 'requires scipy')
+class TestCooMatrixInvalidInit(unittest.TestCase):
+
```
Comment (takagi) [inline tests/cupy_tests/sparse_tests/test_coo.py]: Would you add a test for `shape=None`?

File: tests/cupy_tests/sparse_tests/test_coo.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+    # 1, 2, 0, 0
+    # 0, 0, 0, 3
+    # 0, 0, 4, 0
+    return sp.coo_matrix((data, (row, col)), shape=(3, 4))
+
+
+@testing.parameterize(*testing.product({
+    'dtype': [numpy.float32, numpy.float64],
+}))
+class TestCooMatrix(unittest.TestCase):
```
Comment (takagi) [inline tests/cupy_tests/sparse_tests/test_coo.py]: Would you add tests for `dtype`, `data`, `row` and `col` attributes, including how `copy` parameter works?

File: tests/cupy_tests/sparse_tests/test_coo.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+            [0, 0, 3, 0]
+        ]
+        cupy.testing.assert_allclose(m, expect)
+
+
+@testing.parameterize(*testing.product({
+    'dtype': [numpy.float32, numpy.float64],
+}))
+@unittest.skipUnless(scipy_available, 'requires scipy')
+class TestCooMatrixInvalidInit(unittest.TestCase):
```
Comment (takagi) [inline tests/cupy_tests/sparse_tests/test_coo.py]: Would you add tests for the case that row index and col index exceed matrix dimensions?

Comment (takagi) [inline tests/cupy_tests/sparse_tests/test_coo.py]: Would you add tests for the case that row index and col index are negative?

File: cupy/cusparse.py [hunk 10]
```diff
@@ -117,6 +117,36 @@ def cscsort(x):
...
 
+def coosort(x):
+    handle = device.get_cusparse_handle()
+    m, n = x.shape
+    nnz = x.nnz
+
+    buffer_size = cusparse.xcoosort_bufferSizeExt(
+        handle, m, n, nnz, x.row.data.ptr, x.col.data.ptr)
+    buf = cupy.empty(buffer_size, 'b')
+    P = cupy.empty(nnz, 'i')
```
Comment (unnonouno) [inline cupy/cusparse.py]: Based on cusparse document, it must be `int` not be `size_t`  http://docs.nvidia.com/cuda/cusparse/index.html#cusparse-lt-t-gt-identity
http://docs.nvidia.com/cuda/cusparse/index.html#cusparse-lt-t-gt-coosort

File: tests/cupy_tests/sparse_tests/test_coo.py (hunk 8 above)
Comment (unnonouno) [inline tests/cupy_tests/sparse_tests/test_coo.py]: done

File: tests/cupy_tests/sparse_tests/test_coo.py (hunk 9 above)
Comment (unnonouno) [inline tests/cupy_tests/sparse_tests/test_coo.py]: done

File: tests/cupy_tests/sparse_tests/test_coo.py (hunk 7 above)
Comment (unnonouno) [inline tests/cupy_tests/sparse_tests/test_coo.py]: done

**Comment by unnonouno** [comment]
I updated the test.

File: cupy/cusparse.py (hunk 10 above)
Comment (takagi) [inline cupy/cusparse.py]: Okay, it comes from cusparse spec.

**Comment by takagi** [comment]
LGTM!
